test/HpackSpec.hs

```diff
@@ -94,3 +96,64 @@ spec = do
               old <- readFile file
               hpack `shouldReturn` outputUnchanged
               readFile file `shouldReturn` old
+
+    context "with an absolute path to local defaults" $ do
+      it "resolves the path correctly" $ do
+        withTempDirectory $ flip withCurrentDirectory $ do
+          withTempDirectory $ \ defaultsDir -> do
+            writeFile (defaultsDir </> "defaults.yaml") "ghc-options: A"
+            writeFile "package.yaml" $ unlines
+              [ "name: foo"
+              , "defaults:"
+              , "  local: " ++ (defaultsDir </> "defaults.yaml")
+              , "executable:"
+              , "  main: Main.hs"
+              ]
+            _ <- hpackResult $ setTarget "package.yaml" defaultOptions
+            res <- readFile "foo.cabal"
+            res `shouldContain` "ghc-options: A"
+
+    context "with a relative path to local defaults" $ do
+      it "resolves relative to package.yaml, not CWD" $ do
+        withTempDirectory $ flip withCurrentDirectory $ do
+          createDirectory "foo"
+          writeFile "defaults.yaml" "ghc-options: A"
+          writeFile "foo/defaults.yaml" "ghc-options: B"
+          writeFile "foo/package.yaml" $ unlines
+            [ "name: foo"
+            , "defaults:"
+            , "  local: defaults.yaml"
+            , "executable:"
+            , "  main: Main.hs"
+            ]
+          _ <- hpackResult $ setTarget "foo/package.yaml" defaultOptions
+          fromOutside <- readFile "foo/foo.cabal"
+          _ <- withCurrentDirectory "foo" $ hpackResult defaultOptions
+          readFile "foo/foo.cabal" `shouldReturn` fromOutside
+
+      it "does not have false positives in cycle detection due to same-named default files in different directories" $ do
+        withTempDirectory $ flip withCurrentDirectory $ do
```

Comment:
Hmm, yes, that would be bad, but I think the code would actually get more complicated to implement this bug, say the simpler code gives us the correct behavior.

Usually I refrain from exhaustive testing and only write tests that are necessary to drive the implementation.